cpp/src/fil/infer.cu

```diff
@@ -576,21 +576,53 @@ __global__ void infer_k(storage_type forest, predict_params params) {
   }
 }
 
+void set_carveout(void* kernel, int footprint, int max_shm) {
+  CUDA_CHECK(
+    cudaFuncSetAttribute(kernel, cudaFuncAttributePreferredSharedMemoryCarveout,
+                         // footprint in % of max_shm, rounding up
+                         (100 * footprint + max_shm - 1) / max_shm));
+  CUDA_CHECK(cudaFuncSetAttribute(
+    kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, footprint));
+}
+
+template <int N, leaf_algo_t l, bool c>
+void set_carveouts(int footprint) {
+  int device = 0;
+  CUDA_CHECK(cudaGetDevice(&device));
+  int max_shm = 0;
+  CUDA_CHECK(cudaDeviceGetAttribute(
+    &max_shm, cudaDevAttrMaxSharedMemoryPerBlockOptin, device));
+  if (footprint > max_shm) return;
+
+  set_carveout((void*)infer_k<N, l, c, dense_storage>, footprint, max_shm);
```

Comment:
Do we need to set it for all 3 instantiations? Can't it be set only for the kernel used in this particular model?